### PR TITLE
[webui] KIWI Repositories editor

### PR DESCRIPTION
**Pull request by eduardoj**
Mobprogrammed by: @Ana06, @eduardoj, @mdeniz and @DavidKang

### CODE REVIEW

**Comment by eduardoj** [comment]
[(5) P1: Repository editor part 2: Edit and save repository related data](https://trello.com/c/FB3Gcy55/403-5-p1-repository-editor-part-2-edit-and-save-repository-related-data)

File: src/api/app/controllers/webui/kiwi/images_controller.rb [hunk 1]
```diff
@@ -24,6 +26,20 @@ def import_from_package
 
         redirect_to kiwi_image_repositories_path(package.kiwi_image)
       end
+
+      def outdated
```
Comment (evanrolfe) [inline src/api/app/controllers/webui/kiwi/images_controller.rb]: This could be changed to `def show` and then if we need to render more field from the image then we can add it to the json response.

**Comment by eduardoj** [comment]
@evanrolfe : We renamed `outated` to `show`. Were you also referring to `bulk_edit` and `bulk_upate` actions? What could be the REST approach for this?

**Comment by evanrolfe** [comment]
@eduardoj thanks! TBH I'm not completely sure about the `bulk_edit` or `bulk_update`. I guess having something like `Kiwi::Repositories::Bulk#create` and `#update` could also work but I'm fine if we leave that controller as is.

**Comment by evanrolfe** [comment]
Actually why not just call the methods `edit` and `update`? Since they still only work on a single image.

**Comment by bgeuken** [comment]
and I thought I've been there to

File: src/api/app/controllers/webui/kiwi/repositories_controller.rb [hunk 2]
```diff
@@ -1,12 +1,46 @@
...
         @project = @package.project
       end
+
+      def bulk_edit
+        @repositories = @image.repositories.order(:order)
+        @package = @image.package
+        @project = @package.project
+      end
+
+      def bulk_update
```
Comment (evanrolfe) [inline src/api/app/controllers/webui/kiwi/repositories_controller.rb]: Can we just call these methods `edit` and `delete`? Even though it accepts nested attributes for multiple repositories, look at this rails casts example, he still follows the usual naming conventions:

https://github.com/ryanb/railscasts-episodes/blob/master/episode-197/surveysays/app/controllers/surveys_controller.rb

http://railscasts.com/episodes/197-nested-model-form-part-2

Comment (eduardoj) [inline src/api/app/controllers/webui/kiwi/repositories_controller.rb]: Please, @evanrolfe, review it again.

File: src/api/app/assets/stylesheets/webui/application/kiwi_image.scss [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#update_form {
```
Comment (hennevogel) [inline src/api/app/assets/stylesheets/webui/application/kiwi_image.scss]: This would be a bit too generic for my taste. And you do not style the form, only `.repository-container` and children.

File: src/api/app/views/webui/kiwi/repositories/edit.html.haml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      success: function(json) {
+        var is_outdated = json['is_outdated'];
+        if (is_outdated && !confirm("This repositories have been modified meanwhile you were editing.\nDo you want to apply the changes anyway?"))
+          return;
+        $('#update_form').submit();
+      }
+    });
+  }
+
+  $('#update_form_save').click(save_repositories);
```
Comment (hennevogel) [inline src/api/app/views/webui/kiwi/repositories/edit.html.haml]: Neat but very shaky....

The only thing that keeps you from accidentally updating the kiwi file is this javascript. Which can go wrong for a million reasons.

I would propose you keep this but also make this a validation.

**Comment by hennevogel** [comment]
oh and @eduardoj why shouldn't we merge this?

**Comment by eduardoj** [comment]
@hennevogel I'm still working on completing the tests of the models. Tests of the controllers are done.

File: src/api/app/views/webui/kiwi/repositories/edit.html.haml (hunk 4 above)
Comment (eduardoj) [inline src/api/app/views/webui/kiwi/repositories/edit.html.haml]: Validation added, with a new commit.

File: src/api/app/assets/stylesheets/webui/application/kiwi_image.scss (hunk 3 above)
Comment (eduardoj) [inline src/api/app/assets/stylesheets/webui/application/kiwi_image.scss]: @hennevogel 

> This would be a bit too generic for my taste. And you do not style the form, only .repository-container and children.

Changed `update_form` by `kiwi_repository_update_form`, and changed `.repository-container` by `.kiwi-repository-container`. Better?

File: src/api/spec/models/kiwi/image_spec.rb [hunk 5]
```diff
@@ -76,4 +82,60 @@
...
+      it { expect(subject).to be_truthy }
+    end
+
+    context 'with exception' do
+      before do
+        allow_any_instance_of(Kiwi::Image).to receive(:save!).and_raise(APIException)
+      end
+
+      it { expect{ package.kiwi_image.write_to_backend }.to raise_error(APIException) }
+    end
```
Comment (hennevogel) [inline src/api/spec/models/kiwi/image_spec.rb]: What exactly are you testing in the last two `context`?

Comment (eduardoj) [inline src/api/spec/models/kiwi/image_spec.rb]: @hennevogel 

> What exactly are you testing in the last two `context`?

Short answer: `write_to_backend` method in Kiwi::Image model.

In the first `context` I will add checks for md5 fields, and to assure the file is saved in the backend.

In the second `context`, that it throws an error when calling `save!`. I will leave at it is.

Comment (hennevogel) [inline src/api/spec/models/kiwi/image_spec.rb]: I think what you are testing in the first `context` is that `package.save_file` works (which is already tested somewhere else I suppose) and what you are testing in the second `context` is that `rescue` works which is already tested in the rails test suite ;-)

File: src/api/app/models/kiwi/image.rb [hunk 6]
```diff
@@ -51,6 +53,31 @@ def self.build_from_xml(xml_string, md5)
...
+
+    # Reparser for pretty printing
+    Nokogiri::XML(doc.to_xml, &:noblanks).to_xml
+  end
+
+  def write_to_backend
+    package.save_file({ filename: package.kiwi_image_file, file: to_xml })
+    self.md5_last_revision = package.kiwi_file_md5
+    save!
+  rescue => e
```
Comment (hennevogel) [inline src/api/app/models/kiwi/image.rb]: What's that?

Comment (eduardoj) [inline src/api/app/models/kiwi/image.rb]: @hennevogel : I agree on droping the tests of write_to_backend method

Comment (eduardoj) [inline src/api/app/models/kiwi/image.rb]: @hennevogel : I agree on droping the `rescue => e`. The error is managed in the controller.

File: src/api/app/assets/stylesheets/webui/application/kiwi_image.scss [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+          width: 31%;
+
+          input { width: 96%; }
+          select {
+            width: 100%;
+          }
+        }
+      }
+      &.single, &.double, &.triple {
+        margin-bottom: 0.5em;
```
Comment (Ana06) [inline src/api/app/assets/stylesheets/webui/application/kiwi_image.scss]: for consistency, write this in one single line

File: src/api/spec/controllers/webui/kiwi/repositories_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+# WARNING: Some tests require real backend answers, so make sure you uncomment
+# this line and start a test backend.
+# CONFIG['global_write_through'] = true
+
+RSpec.describe Webui::Kiwi::RepositoriesController, type: :controller, vcr: true do
+  let(:user) { create(:confirmed_user, login: 'tom') }
+  let(:project) { create(:project, name: 'fake_project') }
+  let(:kiwi_image) { create(:kiwi_image) }
+  let(:package) { create(:package_with_kiwi_file, name: 'fake_package', project: project, kiwi_image_id: kiwi_image.id) }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/repositories_controller_spec.rb]: use `kiwi_image` instead `kiwi_image_id`:

``` Ruby
let(:package) { create(:package_with_kiwi_file, name: 'fake_package', project: project, kiwi_image: kiwi_image) }
```

File: src/api/spec/models/kiwi/repository_spec.rb [hunk 9]
```diff
@@ -67,4 +67,19 @@
     it { is_expected.to allow_value(nil).for(:imageinclude) }
     it { is_expected.to allow_value(nil).for(:prefer_license) }
   end
+
+  context 'do not update with outdated image' do
+    let(:kiwi_repository) { create(:kiwi_repository_with_package) }
+    before do
+      allow_any_instance_of(Package).to receive(:kiwi_image_outdated?) { true }
```
Comment (Ana06) [inline src/api/spec/models/kiwi/repository_spec.rb]: that means that every kiwi images needs a Package to get outdated, and I think that was not the original idea. So either we fix this, or we enforce that all kiwi image has a package (introducing a validation). We can do it in a different PR, but we should solve it.

@eduardoj @mdeniz @DavidKang ^^

File: src/api/app/views/webui/kiwi/repositories/_repository_entry.html.haml [hunk 10]
```diff
@@ -5,10 +5,10 @@
     %i (#{repo_type_and_priority(repository)})
   %div{ style: "margin-left: 30px" }
     %div{ style: "margin: 5px" }
-      - if repository.username
+      - if repository.username.present?
```
Comment (eduardoj) [inline src/api/app/views/webui/kiwi/repositories/_repository_entry.html.haml]: Done. Now it is in a different commit.

File: src/api/app/views/webui/kiwi/repositories/_repository_entry.html.haml [hunk 11]
```diff
@@ -5,10 +5,10 @@
...
         %strong User:
         = repository.username
-      - if repository.password
+      - if repository.password.present?
```
Comment (eduardoj) [inline src/api/app/views/webui/kiwi/repositories/_repository_entry.html.haml]: Done. Now it is in a different commit.

File: src/api/app/models/kiwi/image.rb [hunk 12]
```diff
@@ -51,6 +53,29 @@ def self.build_from_xml(xml_string, md5)
...
+    image = doc.at_css('image')
+    image.first_element_child.after(xml_repos)
+
+    # Reparser for pretty printing
+    Nokogiri::XML(doc.to_xml, &:noblanks).to_xml
+  end
+
+  def write_to_backend
+    package.save_file({ filename: package.kiwi_image_file, file: to_xml })
+    self.md5_last_revision = package.kiwi_file_md5
```
Comment (bgeuken) [inline src/api/app/models/kiwi/image.rb]: I didn't find any check for the kiwi file md5sum in ``` save_file! ```. This should be checked before doing any writes. And especially before updating the ``` self.md5_last_revision ```

File: src/api/app/models/kiwi/repository.rb [hunk 13]
```diff
@@ -50,7 +53,31 @@ def source_path_format
...
+
+    builder.to_xml save_with: Nokogiri::XML::Node::SaveOptions::NO_DECLARATION | Nokogiri::XML::Node::SaveOptions::FORMAT
+  end
+
   #### Alias of methods
+
+  private
+
+  def not_outdated
+    errors.add(:base, 'Image configuration has changed') if image.package.kiwi_image_outdated?
```
Comment (bgeuken) [inline src/api/app/models/kiwi/repository.rb]: btw, the kiwi_image_outdated? method is missing tests:-) I know it's not added here though

File: src/api/app/models/kiwi/image.rb (hunk 12 above)
Comment (eduardoj) [inline src/api/app/models/kiwi/image.rb]: The check of the kiwi file md5sum is done in the frontend, in the view. Have a look [here](https://github.com/openSUSE/open-build-service/pull/3225/files#diff-030dd5bfde342910fc0161df227de456R15)

As the form button is clicked, a JSON request is trown, and a "is_outdated" state is expected:
- If "is_outdated" is true, a confirm dialog is thrown, asking for the confirmation of the user to apply the changes anyway (different md5sum).
- If "is_outdated" is false, the update action is requested (same md5sum).

File: src/api/spec/models/kiwi/repository_spec.rb (hunk 9 above)
Comment (eduardoj) [inline src/api/spec/models/kiwi/repository_spec.rb]: > @Ana06: that means that every kiwi images needs a Package to get outdated, and I think that was not the original idea. So either we fix this, or we enforce that all kiwi image has a package (introducing a validation). We can do it in a different PR, but we should solve it.
> 
> @eduardoj @mdeniz @DavidKang ^^
> 

We'll fix this in a different PR.

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+# this line and start a test backend.
+# CONFIG['global_write_through'] = true
+
+RSpec.describe Webui::Kiwi::ImagesController, type: :controller, vcr: true do
+  let(:project) { create(:project, name: 'fake_project') }
+  let(:kiwi_image) { create(:kiwi_image) }
+
+  describe 'GET #import_from_package' do
+    context 'without a kiwi file' do
+      let(:package) { create(:package, name: 'package_without_kiwi_file', project: project, kiwi_image: kiwi_image) }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: `package` is to generic. What about `package_with_kiwi_image`?

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+      it { expect(response).to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq('There is no KIWI file') }
+    end
+
+    context 'with a kiwi file' do
+      context 'with a valid kiwi file' do
+        include_context 'a kiwi image xml'
+
+        let(:package) do
```
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: `package` is to generic. What about `package_with_kiwi_file`?

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+          package.reload
+        end
+
+        it { expect(response).to redirect_to(kiwi_image_repositories_path(package.kiwi_image)) }
+      end
+
+      context 'with a invalid kiwi file' do
+        include_context 'an invalid kiwi image xml'
+
+        let(:package) do
```
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: `package` is to generic. What about `package_with_kiwi_image`?

File: src/api/spec/controllers/webui/kiwi/repositories_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+# WARNING: Some tests require real backend answers, so make sure you uncomment
+# this line and start a test backend.
+# CONFIG['global_write_through'] = true
+
+RSpec.describe Webui::Kiwi::RepositoriesController, type: :controller, vcr: true do
+  let(:user) { create(:confirmed_user, login: 'tom') }
+  let(:project) { create(:project, name: 'fake_project') }
+  let(:kiwi_image) { create(:kiwi_image) }
+  let(:package) { create(:package_with_kiwi_file, name: 'fake_package', project: project, kiwi_image: kiwi_image) }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/repositories_controller_spec.rb]: `package` is to generic. What about `package_with_kiwi_file`?

File: src/api/spec/factories/kiwi_image.rb [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    name { Faker::Name.name }
+    md5_last_revision { Faker::Crypto.md5 }
+
+    factory :kiwi_image_with_package do
+      transient do
+        package_name { 'package_with_kiwi_image' }
+      end
+
+      after(:create) do |image, evaluator|
+        image.package = create(:package, name: evaluator.package_name)
```
Comment (Ana06) [inline src/api/spec/factories/kiwi_image.rb]: this two lines cna be changed by:

``` Ruby
create(:package, name: evaluator.package_name, kiwi_image: image)
```

as you don't need to save it if you write it like that

File: src/api/spec/factories/kiwi_repository.rb [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+    association :image, factory: :kiwi_image
+
+    source_path 'http://example.com/'
+    order 1
+    repo_type { Kiwi::Repository::REPO_TYPES.first }
+    replaceable false
+
+    factory :kiwi_repository_with_package do
+      after(:create) do |repository|
+        repository.image.package = create(:package)
```
Comment (Ana06) [inline src/api/spec/factories/kiwi_repository.rb]: same here: 

``` Ruby
create(:package, kiwi_image: repository.image.)
```

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb (hunk 14 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: and what about a factory for `package_with_kiwi_image`?

File: src/api/spec/factories/kiwi_image.rb (hunk 18 above)
Comment (eduardoj) [inline src/api/spec/factories/kiwi_image.rb]: It will be also needed a `image.reload`. @Ana06 and I decided to leave it as it is.

File: src/api/spec/factories/kiwi_repository.rb (hunk 19 above)
Comment (eduardoj) [inline src/api/spec/factories/kiwi_repository.rb]: It will be also needed a `repository.reload`. @Ana06 and I decided to leave it as it is.

**Comment by hennevogel** [comment]
@Ana06 please re-review this.

**Comment by eduardoj** [comment]
@Ana06 : Your changes are already applied.

File: src/api/app/models/kiwi/repository.rb [hunk 20]
```diff
@@ -23,7 +23,7 @@ class Kiwi::Repository < ApplicationRecord
   validates :order, numericality: { only_integer: true, greater_than_or_equal_to: 1 }
   # TODO: repo_type value depends on packagemanager element
   # https://doc.opensuse.org/projects/kiwi/doc/#sec.description.repository
-  validates :repo_type, inclusion: { in: %w(apt-deb rpm-dir rpm-md yast2) }
+  validates :repo_type, inclusion: { in: %w[apt-deb rpm-dir rpm-md yast2] }
```
Comment (Ana06) [inline src/api/app/models/kiwi/repository.rb]: :question: when was this offense introduced? This validation was already there and in master Rubocop is working :thinking:

File: src/api/app/models/kiwi/repository.rb [hunk 21]
```diff
@@ -22,9 +23,10 @@ class Kiwi::Repository < ApplicationRecord
...
```
Comment (Ana06) [inline src/api/app/models/kiwi/repository.rb]: :question: when was this offense introduced? This validation was already there and in master Rubocop is working :thinking:

File: src/api/spec/models/kiwi/image_spec.rb [hunk 22]
```diff
@@ -76,4 +84,159 @@
...
+      end
+
+      subject { create(:kiwi_image_with_package, project: project, with_kiwi_file: true, file_content: 'Invalid content for a xml file') }
+
+      it { expect(subject.to_xml).to be_nil }
+    end
+
+    context 'with a invalid kiwi image file (without image children)' do
+      after do
+        # We need to be logged to destroy a package
```
Comment (Ana06) [inline src/api/spec/models/kiwi/image_spec.rb]: I think this is obviously, as you are just login the user before destroying the package, and logging it out just after that. And this is also something really normal in our code. I would like to remove this comment.

File: src/api/spec/models/kiwi/image_spec.rb [hunk 23]
```diff
@@ -76,4 +84,159 @@
...
+      it { expect(subject.errors).to be_empty }
+      it { expect(subject.xpath('.//image').length).to be(1) }
+      it { expect(subject.xpath('.//image/description').length).to be(1) }
+      it { expect(subject.xpath('.//image/packages/package').length).to be(20) }
+      it { expect(subject.xpath('.//image/repository').length).to be(0) }
+    end
+
+    context 'with a invalid kiwi image file' do
+      after do
+        # We need to be logged to destroy a package
```
Comment (Ana06) [inline src/api/spec/models/kiwi/image_spec.rb]: I think this is obviously, as you are just login the user before destroying the package, and logging it out just after that. And this is also something really normal in our code. I would like to remove this comment.

File: src/api/spec/models/kiwi/image_spec.rb [hunk 24]
```diff
@@ -76,4 +84,159 @@
...
+        it { expect(subject.xpath('.//image').length).to be(1) }
+        it { expect(subject.xpath('.//image/description').length).to be(1) }
+        it { expect(subject.xpath('.//image/packages').length).to be(1) }
+        it { expect(subject.xpath('.//image/repository').length).to be(1) }
+      end
+    end
+
+    context 'without kiwi image file' do
+      after do
+        # We need to be logged to destroy a package
```
Comment (Ana06) [inline src/api/spec/models/kiwi/image_spec.rb]: I think this is obviously, as you are just login the user before destroying the package, and logging it out just after that. And this is also something really normal in our code. I would like to remove this comment.

File: src/api/spec/models/kiwi/image_spec.rb [hunk 25]
```diff
@@ -76,4 +84,159 @@
...
+      it 'returns nil' do
+        expect(subject.to_xml).to be_nil
+      end
+    end
+
+    context 'with kiwi image file' do
+      let(:kiwi_image) { create(:kiwi_image_with_package, project: project, with_kiwi_file: true, file_content: kiwi_xml) }
+
+      after do
+        # We need to be logged to destroy a package
```
Comment (Ana06) [inline src/api/spec/models/kiwi/image_spec.rb]: I think this is obviously, as you are just login the user before destroying the package, and logging it out just after that. And this is also something really normal in our code. I would like to remove this comment.

File: src/api/spec/factories/kiwi_image.rb [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+    name { Faker::Name.name }
+    md5_last_revision nil
+
+    factory :kiwi_image_with_package do
+      transient do
+        package_name 'package_with_kiwi_image'
+        with_kiwi_file false
+        kiwi_file_name { "#{package_name}.kiwi" }
+        file_content { Kiwi::Image::DEFAULT_KIWI_BODY }
+        project nil
```
Comment (Ana06) [inline src/api/spec/factories/kiwi_image.rb]: why is this here? if you remove this line then it would be `nil` anyway, right?

File: src/api/app/controllers/webui/kiwi/images_controller.rb [hunk 27]
```diff
@@ -24,6 +26,17 @@ def import_from_package
...
       end
+
+      def show
+        render json: { is_outdated: @image.outdated? }
+      end
+
+      private
+
+      def set_image
+        required_parameters :id
```
Comment (evanrolfe) [inline src/api/app/controllers/webui/kiwi/images_controller.rb]: Can you use `params.require`?

File: src/api/spec/factories/kiwi_image.rb (hunk 26 above)
Comment (mdeniz) [inline src/api/spec/factories/kiwi_image.rb]: Mmmm.. no I guess, if you don put this line is not declared as a transient param and whenever you use this it will try to assign it to a Kiwi::Image model...

File: src/api/app/controllers/webui/kiwi/repositories_controller.rb [hunk 28]
```diff
@@ -1,12 +1,43 @@
...
+      private
+
+      def image_params
+        params.require(:image).permit(repositories_attributes:
+                                      [:id, :priority, :repo_type, :source_path, :alias,
+                                       :username, :password, :prefer_license, :imageinclude, :replaceable])
+      end
+
+      def set_image
+        required_parameters :kiwi_image_id
```
Comment (evanrolfe) [inline src/api/app/controllers/webui/kiwi/repositories_controller.rb]: Can you use params.require?

File: src/api/spec/factories/kiwi_image.rb (hunk 26 above)
Comment (Ana06) [inline src/api/spec/factories/kiwi_image.rb]: right! :wink:

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+        get :import_from_package, params: { package_id: package.id }
+      end
+
+      it { expect(response).to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq('There is no KIWI file') }
+    end
+
+    context 'with a kiwi file' do
+      context 'with a valid kiwi file' do
+        let(:kiwi_image_with_package_with_kiwi_file) do
```
Comment (evanrolfe) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: [Nitpick] you can shorten the variable name to just `kiwi_image` and then expand the context to:
`context 'with a valid kiwi file with a package`

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+          create(:kiwi_image_with_package,
+                 name: 'package_with_valid_kiwi_file', project: project, with_kiwi_file: true)
+        end
+
+        before do
+          get :import_from_package, params: { package_id: kiwi_image_with_package_with_kiwi_file.package.id }
+          kiwi_image_with_package_with_kiwi_file.reload
```
Comment (evanrolfe) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: Why is this necessary?

File: src/api/spec/models/kiwi/image_spec.rb [hunk 31]
```diff
@@ -76,4 +84,159 @@
...
+        it { expect(subject.xpath('.//image/packages').length).to be(1) }
+        it { expect(subject.xpath('.//image/repository').length).to be(1) }
+      end
+    end
+
+    context 'without kiwi image file' do
+      after do
+        # We need to be logged to destroy a package
+        login user
+        subject.package.destroy
```
Comment (evanrolfe) [inline src/api/spec/models/kiwi/image_spec.rb]: Why does the package need to be destroyed? Don't the rspec tests run in transactions which get rolled back once the test completes?

File: src/api/spec/models/kiwi/image_spec.rb [hunk 32]
```diff
@@ -76,4 +84,159 @@
       it { expect(subject.valid?).to be_falsey }
     end
   end
+
+  describe '.to_xml' do
```
Comment (evanrolfe) [inline src/api/spec/models/kiwi/image_spec.rb]: Should be `#to_xml` since its an instance method:

http://www.betterspecs.org/#describe

File: src/api/spec/models/kiwi/image_spec.rb (hunk 25 above)
Comment (bgeuken) [inline src/api/spec/models/kiwi/image_spec.rb]: Are you sure it's obvious?

File: src/api/spec/models/kiwi/image_spec.rb (hunk 23 above)
Comment (bgeuken) [inline src/api/spec/models/kiwi/image_spec.rb]: You are repeating yourself :laughing:

File: src/api/spec/models/kiwi/image_spec.rb (hunk 22 above)
Comment (bgeuken) [inline src/api/spec/models/kiwi/image_spec.rb]: You are repeating yours... oh my:-)

**Comment by eduardoj** [comment]
@Ana:
> @eduardoj please drop 9fccae8 it is not needed

Done. And all the other requests too.

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb (hunk 30 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: @evanrolfe 
> Why is this necessary?

It wasn't. Removed.

File: src/api/spec/models/kiwi/image_spec.rb (hunk 31 above)
Comment (mdeniz) [inline src/api/spec/models/kiwi/image_spec.rb]: Transactions only work in the DB... not in the backend 😞 , so the file created needs to be deleted too.

Comment (evanrolfe) [inline src/api/spec/models/kiwi/image_spec.rb]: Ahhh, then if this is only needed for the backend, maybe we should do this?
```ruby
if CONFIG['global_write_through']
  login user
  subject.package.destroy
end
```

Comment (mdeniz) [inline src/api/spec/models/kiwi/image_spec.rb]: Why? This is done already at destroying the package inside `write_to_backend`...
